java/ql/test/query-tests/security/CWE-089/semmle/examples/AllowListSanitizerWithJavaUtilSet.java

```diff
@@ -0,0 +1,305 @@
+// Test cases for CWE-089 (SQL injection and Java Persistence query injection)
+// http://cwe.mitre.org/data/definitions/89.html
+package test.cwe089.semmle.tests;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+import java.util.function.Consumer;
+
+class AllowListSanitizerWithJavaUtilSet {
+	public static Connection connection;
+	public static final Set<String> goodAllowList1 = Set.of("allowed1", "allowed2", "allowed3");
+	public static final Set<String> goodAllowList2 = Collections.unmodifiableSet(new HashSet<String>(Arrays.asList("allowed1","allowed2")));
+	public static final Set<String> goodAllowList3;
+	public static final Set<String> goodAllowList4;
+	public static final Set<String> badAllowList1 = Set.of("allowed1", "allowed2", getNonConstantString());
+	public static final Set<String> badAllowList2 = Collections.unmodifiableSet(new HashSet<String>(Arrays.asList("allowed1", getNonConstantString())));
+	public static final Set<String> badAllowList3;
+	public static final Set<String> badAllowList4;
+	public static final Set<String> badAllowList5;
+	public static Set<String> badAllowList6 = Set.of("allowed1", "allowed2", "allowed3");
+	public final Set<String> badAllowList7 = Set.of("allowed1", "allowed2", "allowed3");
```

Comment:
I wrote these as negative examples because it's hard to be sure that they don't contain a non-constant element. In practice, when people want to check a value is in an allowed list of constants, I think they construct that list locally if it will be used once or put it in a static field if it might be used more broadly, and they generally make that list final and immutable, so that it's clear that it only has the contents that it seems to have on first glance.

Reply:
In other words, I think we should err on the side of being cautious when identifying these sanitizers, and I think it's reasonable to be able to say to a user who complains about an FP "we didn't recognize that sanitizer, but if you made it a static final field/immutable then we would."